packages/kbn-dev-utils/src/ci_stats_reporter/ci_stats_reporter.ts

```diff
@@ -7,106 +7,187 @@
+    const defaultMetadata = {
+      osPlatform: Os.platform(),
+      osRelease: Os.release(),
+      osArch: Os.arch(),
+      cpuCount: Os.cpus()?.length,
+      cpuModel: Os.cpus()[0]?.model,
+      cpuSpeed: Os.cpus()[0]?.speed,
+      kibanaUuid,
+    };
+
+    return await this.req({
+      auth: !!buildId,
+      path: '/v1/timings',
+      body: {
+        buildId,
+        upstreamBranch,
+        timings,
+        defaultMetadata,
+      },
+      bodyDesc: timings.length === 1 ? `${timings.length} timing` : `${timings.length} timings`,
+    });
   }
 
-  return config as Config;
-}
+  /**
+   * Report metrics data to the ci-stats service. If running outside of CI this method
+   * does nothing as metrics can only be reported when associated with a specific CI build.
+   */
+  async metrics(metrics: CiStatsMetric[]) {
+    if (!this.hasBuildConfig()) {
+      return;
+    }
 
-export class CiStatsReporter {
-  static fromEnv(log: ToolingLog) {
-    return new CiStatsReporter(parseConfig(log), log);
-  }
+    const buildId = this.config?.buildId;
 
-  constructor(private config: Config | undefined, private log: ToolingLog) {}
+    if (!buildId) {
+      throw new Error(`CiStatsReporter can't be authorized without a buildId`);
+    }
 
-  isEnabled() {
-    return !!this.config;
+    return await this.req({
+      auth: true,
+      path: '/v1/metrics',
+      body: {
+        buildId,
+        metrics,
+      },
+      bodyDesc: `metrics: ${metrics
+        .map(({ group, id, value }) => `[${group}/${id}=${value}]`)
+        .join(' ')}`,
+    });
   }
 
-  async metrics(metrics: CiStatsMetrics) {
-    if (!this.config) {
-      return;
+  /**
+   * In order to allow this code to run before @kbn/utils is built, @kbn/pm will pass
+   * in the upstreamBranch when calling the timings() method. Outside of @kbn/pm
+   * we rely on @kbn/utils to find the package.json file.
+   */
+  private getUpstreamBranch() {
+    // specify the module id in a way that will keep webpack from bundling extra code into @kbn/pm
+    const hideFromWebpack = ['@', 'kbn/utils'];
+    // eslint-disable-next-line @typescript-eslint/no-var-requires
+    const { kibanaPackageJson } = require(hideFromWebpack.join(''));
+    return kibanaPackageJson.branch;
+  }
+
+  /**
+   * In order to allow this code to run before @kbn/utils is built, @kbn/pm will pass
+   * in the kibanaUuid when calling the timings() method. Outside of @kbn/pm
+   * we rely on @kbn/utils to find the repo root.
+   */
+  private getKibanaUuid() {
+    // specify the module id in a way that will keep webpack from bundling extra code into @kbn/pm
+    const hideFromWebpack = ['@', 'kbn/utils'];
+    // eslint-disable-next-line @typescript-eslint/no-var-requires
+    const { REPO_ROOT } = require(hideFromWebpack.join(''));
+    try {
+      return Fs.readFileSync(Path.resolve(REPO_ROOT, 'data/uuid'), 'utf-8').trim();
```

Comment:
It's probably not worth accounting for as it's not common to change in dev, but the data directory can be configured with `path.data`.

Reply:
Yeah, in that case I think it's fine to ignore the UUID.